### PR TITLE
refactor Operator::close,set_finishing,set_finished,set_canceled

### PR DESCRIPTION
**Pull request by satanson**
## What type of PR is this：
- [ ] bug
- [ ] feature
- [x] enhancement
- [ ] others

## Problem Summary(Required) ：
Operator::{close, set_finishing, set_finished, set_canceled} return Status instead void, because some operators (e.g. HashJoin)  work in full-materialization style use set_finishing function to complete its final work after receiving all chunks. so set_finishing function maybe fail because of memory allocation failure (throw std::bad_alloc and mem usage exceeds limit).

### CODE REVIEW

File: be/src/exec/pipeline/pipeline_driver.cpp [hunk 1]
```diff
@@ -392,72 +392,73 @@ bool PipelineDriver::_check_fragment_is_canceled(RuntimeState* runtime_state) {
...
+        return Status::OK();
     }
 
     VLOG_ROW << strings::Substitute("[Driver] close operator [driver=$0] [operator=$1]", to_readable_string(),
                                     op->get_name());
     {
         SCOPED_TIMER(op->_close_timer);
-        op->close(state);
+        op_state = OperatorStage::CLOSED;
+        RETURN_IF_ERROR(op->close(state));
```
Comment (ZiheLiu) [inline be/src/exec/pipeline/pipeline_driver.cpp:456]: If `op->close` returns error, `close()` of the operators after `op` cannot be called. Will this cause memory leak?

File: be/src/exec/pipeline/aggregate/aggregate_blocking_sink_operator.cpp [hunk 2]
```diff
@@ -12,12 +12,12 @@ Status AggregateBlockingSinkOperator::prepare(RuntimeState* state) {
     return _aggregator->open(state);
 }
 
-void AggregateBlockingSinkOperator::close(RuntimeState* state) {
+Status AggregateBlockingSinkOperator::close(RuntimeState* state) {
     _aggregator->unref(state);
-    Operator::close(state);
+    return Operator::close(state);
```
Comment (stdpain) [inline be/src/exec/pipeline/aggregate/aggregate_blocking_sink_operator.cpp:17]: not suggest add Status to close.
if we write such code. may be cause a leak.
close method should be exception safe.
```
RETURN_IF_ERROR(child->close());
RETURN_IF_ERROR(this->close());
```

Comment (satanson) [reply be/src/exec/pipeline/aggregate/aggregate_blocking_sink_operator.cpp:17]: Operator::close(state) is always the last statement of its derived classes' close method

File: be/src/runtime/current_thread.h [hunk 3]
```diff
@@ -192,14 +192,14 @@ class CurrentThreadCheckMemLimitSetter {
...
 };
-
 #define TRY_CATCH_BAD_ALLOC(stmt)                                                           \
     do {                                                                                    \
         try {                                                                               \
             bool prev = tls_thread_status.set_is_catched(true);                             \
             DeferOp op([&] { tls_thread_status.set_is_catched(prev); });                    \
             { stmt; }                                                                       \
         } catch (std::bad_alloc const&) {                                                   \
+            LOG(WARNING) << "Throw std::bad_alloc";                                         \
```
Comment (trueeyu) [inline be/src/runtime/current_thread.h:202]: remove the log

**Review by murphyatwork — DISMISSED** [review]

**Review by stdpain — APPROVED** [review]

**Review by kangkaisen — APPROVED** [review]

**Pull request merged**